### PR TITLE
Refactor/api/generic

### PR DESCRIPTION
**Pull request by Someone-rahul**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -17,67 +17,61 @@ import Rating from "./admin/pages/Rating";
...
+              <Route path="blog" element={<Blog />} />
+              <Route path="about" element={<About />} />
+              <Route path="login" element={<Login />} />
+              <Route path="download" element={<Download />} />
+            </Route>
+            <Route path="/admin/*" element={<Admin_Layout />}>
+              <Route index element={<Dashboard />} />
+              <Route path="onlinedriver" element={<Onlinedriver />} />
+              <Route path="drivers" element={<Driver />} />
+              <Route path="drivers/:id" element={<DriverInfo />} />
```
Comment (ravigrauniyar) [inline src/App.tsx:44]: Please remove this duplicate "drivers/:id" route.

**Review by ravigrauniyar — CHANGES_REQUESTED** [review]

File: src/admin/components/OnlineDriver/OnlineDriverTable.tsx [hunk 2]
```diff
@@ -29,39 +21,33 @@ function OnlineDriverTable() {
     return () => clearInterval(intervalId);
   }, []);
 
-//   if (loadError) {
-//     return <div>Error loading maps</div>;
-//   }
-//   if (!isLoaded) {
-//     return <div>Loading maps...</div>;
-//   }
+  //   if (loadError) {
```
Comment (ravigrauniyar) [inline src/admin/components/OnlineDriver/OnlineDriverTable.tsx:24]: Remove the commented code.

File: src/admin/components/OnlineDriver/OnlineDriverTable.tsx [hunk 3]
```diff
@@ -29,39 +21,33 @@ function OnlineDriverTable() {
...
-                        </tr>
-                    </thead>
-                    <tbody>
-        {Onlinedrivers.map(driver => (
-          <tr key={driver.id}>
-            <td className='py-2 px-4 border-b'>{driver.name}</td>
-            <td className='py-2 px-4 border-b'>{driver.Status}</td>
-            <td className='py-2 px-4 border-b'>{driver.rating}</td>
+    <div className="flex-none h-auto w-1/2 border-2 border-cyan-600 mx-2 my-2">
+      <table className="min-w-full  text-center">
```
Comment (ravigrauniyar) [inline src/admin/components/OnlineDriver/OnlineDriverTable.tsx:32]: Remove the extra space in between class names.

File: src/admin/components/OnlineDriver/OnlineDriverTable.tsx [hunk 4]
```diff
@@ -29,39 +21,33 @@ function OnlineDriverTable() {
...
+        <thead>
+          <tr className="">
```
Comment (ravigrauniyar) [inline src/admin/components/OnlineDriver/OnlineDriverTable.tsx:34]: Please remove the props with no values.

File: src/admin/components/OnlineDriver/OnlineDriverTable.tsx [hunk 5]
```diff
@@ -29,39 +21,33 @@ function OnlineDriverTable() {
...
           </tr>
-        ))}
-      </tbody>
-                </table>
-
-
-            </div>
+        </thead>
+        <tbody>
+          {Onlinedrivers.map((driver) => (
```
Comment (ravigrauniyar) [inline src/admin/components/OnlineDriver/OnlineDriverTable.tsx:41]: Use lodash package for mapping and other array manipulations.

File: src/admin/components/Rides/Ridestable.tsx [hunk 6]
```diff
@@ -1,78 +1,68 @@
...
-
-import { RidesRepo, IRidesRepo } from '../../../repositories/RideRepo';
-import { IRideApi } from '../../../api/type';
-
-import { RidesApi } from '../../../api/RidesApi';
-
+import { useEffect, useState } from "react";
+import { useRepository } from "../../../hooks/CustomHook";
 
 interface Rides {
```
Comment (ravigrauniyar) [inline src/admin/components/Rides/Ridestable.tsx:5]: Move all interfaces or data types to a single "data/types.d.ts" file.

File: src/admin/components/OnlineDriver/OnlineDriverTable.tsx [hunk 7]
```diff
@@ -29,39 +21,33 @@ function OnlineDriverTable() {
     return () => clearInterval(intervalId);
   }, []);
```
Comment (ravigrauniyar) [inline src/admin/components/OnlineDriver/OnlineDriverTable.tsx:22]: Use ESLint extension to identify warnings like adding "repo" to the dependency array "[]" here and in all other similar cases.

File: src/admin/components/Rides/Ridestable.tsx [hunk 8]
```diff
@@ -1,78 +1,68 @@
...
+            <th className="py-2 px-4 border-b">PickupLocation</th>
+            <th className="py-2 px-4 border-b">DropLocation</th>
+            <th className="py-2 px-4 border-b"> distance </th>
+            <th className="py-2 px-4 border-b"> Price</th>
+            <th className="py-2 px-4 border-b">NumberOfPassenger </th>
+            <th className="py-2 px-4 border-b">Message</th>
+          </tr>
+        </thead>
+        <tbody>
+          {Rides.map((Rides) => (
```
Comment (ravigrauniyar) [inline src/admin/components/Rides/Ridestable.tsx:52]: Fix issues similar to those in `OnlineDriverTable`

File: src/admin/components/dashboard/Dashboard_map.tsx [hunk 9]
```diff
@@ -1,9 +1,7 @@
 import { useEffect, useState } from "react";
 import { GoogleMap, useLoadScript, Marker } from "@react-google-maps/api";
-import { IOnlineDriver } from "../../../api/type";
-import { OnlineDriverApi } from "../../../api/OnlineDriverApi";
-import { OnlineDriverRepo } from "../../../repositories/OnlineDriverRepo";
-import logo from '../../../../public/logo.jpg';
+import logo from "../../../../public/logo.jpg";
+import { useRepository } from "../../../hooks/CustomHook";
```
Comment (ravigrauniyar) [inline src/admin/components/dashboard/Dashboard_map.tsx:4]: Let's set up "alias" in config files to shorten the import path to "@hooks/CustomHook".

File: src/admin/components/OnlineDriver/OnlineDriverTable.tsx [hunk 10]
```diff
@@ -1,20 +1,12 @@
...
-
-
+import { useRepository } from "../../../hooks/CustomHook";
 
 function OnlineDriverTable() {
-  
+  const { repo } = useRepository();
   const [Onlinedrivers, setOnlineDrivers] = useState<OnlineDriver[]>([]);
   useEffect(() => {
     const fetchData = async () => {
```
Comment (ravigrauniyar) [inline src/admin/components/OnlineDriver/OnlineDriverTable.tsx:8]: Let's move this `fetchData` method to "actions/fetchOnlineDriverData" or similar and use that method wherever necessary.

File: src/admin/components/driver/DriverInfo.tsx [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+import { useEffect, useState } from "react";
+import { useRepository } from "../../../hooks/CustomHook";
+import { capitalize } from "../../../utils/utilities";
+export default function DriverInfo() {
+  const [open, setOpen] = useState<boolean>(false);
+  const [report, setReport] = useState<string>("");
+  const { id } = useParams();
+  const { repo } = useRepository();
+  function handleSendMessage() {
+    if (confirm("Are you sure!") == true) {
```
Comment (ravigrauniyar) [inline src/admin/components/driver/DriverInfo.tsx:13]: Let's use lodash package's "isEqual" method to check equality whenever necessary.

File: src/admin/components/driver/DriverInfo.tsx [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+    };
+
+    fetchData();
+  }, []);
+
+  // console.log(driver);
+  if (typeof driver === "undefined") {
+    console.log(driver);
+  } else {
+    return (
```
Comment (ravigrauniyar) [inline src/admin/components/driver/DriverInfo.tsx:55]: Let's make it "return !_.isUndefined(driver) && (".

File: src/admin/components/driver/DriverInfo.tsx [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        }
+      } catch (error) {
+        console.error("Error fetching data:", error);
+      }
+    };
+
+    fetchData();
+  }, []);
+
+  // console.log(driver);
```
Comment (ravigrauniyar) [inline src/admin/components/driver/DriverInfo.tsx:51]: Remove the comment and the if ... else condition.

File: src/admin/components/driver/DriverInfo.tsx [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+    return (
+      <>
+        <div className="bg-gray-100 mb-20">
+          <div className="container mx-auto py-2">
+            <div className="flex flex-row gap-x-4">
+              <div className="w-1/2">
+                <div className="bg-white shadow rounded-lg p-6">
+                  <div className="flex flex-col items-center">
+                    <img
+                      src="https://images.unsplash.com/photo-1438761681033-6461ffad8d80?q=80&w=1470&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D"
```
Comment (ravigrauniyar) [inline src/admin/components/driver/DriverInfo.tsx:64]: Let's import this from ".env" file using VITE env variable.

File: src/admin/components/driver/DriverInfo.tsx [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+              </div>
+              <div className="w-full">
+                <div className="bg-white shadow rounded-lg p-6">
+                  <ul className="flex flex-col gap-y-4">
+                    <li>
+                      <h3 className="font-medium text-xl mb-4">
+                        Vehicle Image
+                      </h3>
+                      <img
+                        src="https://images.unsplash.com/photo-1626149637281-4e227308da18?q=80&w=1365&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D"
```
Comment (ravigrauniyar) [inline src/admin/components/driver/DriverInfo.tsx:171]: Same for all src values, use env variables.

File: src/admin/components/driver/DriverTableRow.tsx [hunk 16]
```diff
@@ -41,20 +30,22 @@ export default function DriverTableRow({ user, index }: tableRowProp) {
       <td className="px-4 py-2">{user.ratings}</td>
       <td className={`px-4 py-2 `}>
         <span
-          className={`${user.account_status === ("Not Verified" || "not verified")
-            ? "text-white bg-red-500 py-2 px-4 rounded"
-            : "text-white bg-green-500 py-2 px-4 rounded"
-            }`}
+          className={`${
+            user.account_status === ("Not Verified" || "not verified")
```
Comment (ravigrauniyar) [inline src/admin/components/driver/DriverTableRow.tsx:34]: Use Enum data type for conditional classes.
`enum accountStatusStyles {
    verified= "text-white bg-red-500 py-2 px-4 rounded",
    notVerified= "text-white bg-green-500 py-2 px-4 rounded"
}`

File: src/admin/components/driver/Drivertable.tsx [hunk 17]
```diff
@@ -1,22 +1,33 @@
...
-  if (!driverRepo) {
-    return null;
-  }
-  // eslint-disable-next-line react-hooks/rules-of-hooks
-  const { data, error, isLoading, isError } = useDrivers();
+declare type driverTableProp = {
+  filterValue: string;
+};
+function Drivertable({ filterValue }: driverTableProp) {
+  console.log(filterValue);
```
Comment (ravigrauniyar) [inline src/admin/components/driver/Drivertable.tsx:10]: Please remove all console.log except for logging errors.

File: src/admin/components/shared/Pagination.tsx [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+  //   flex items-center justify-center px-4 h-10 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700 dark:bg-gray-800 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white
+  //   z-10 flex items-center justify-center px-4 h-10 leading-tight text-blue-600 border border-blue-300 bg-blue-50 hover:bg-blue-100 hover:text-blue-700 dark:border-gray-700 dark:bg-gray-700 dark:text-white
+  return (
+    <nav>
+      <ul className="flex items-center -space-x-px h-10 text-base">
+        {pageNumbers.map((number) => (
+          <li key={number}>
+            <button
+              onClick={() => paginate(number)}
+              className="flex items-center justify-center px-4 h-10 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700 dark:bg-gray-800 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white"
```
Comment (ravigrauniyar) [inline src/admin/components/shared/Pagination.tsx:24]: Create variables like "containerStyles", "textStyles", "borderStyles", "darkModeStyles", etc to break down these large styles values.

File: src/api/Api.ts [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+
+  //building the headers
+  private _buildHeaders(headers: Map<string, string>): Headers {
+    const _headers = new Headers();
+    headers.forEach((value, key) => {
+      _headers.append(key, value);
+    });
+    return _headers;
+  }
+  //generate new access token if exipired
```
Comment (ravigrauniyar) [inline src/api/Api.ts:21]: Remove all commented codes.

File: src/client/components/register/DriverRegisterForm.tsx [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+    | "address"
+    | "userImage"
+    | "liscenceNumber"
+    | "liscenceImage"
+    | "billBookImage";
+  value?: string | undefined;
+  type: string;
+  required: boolean;
+};
+const formValue: formValueData[] = [
```
Comment (ravigrauniyar) [inline src/client/components/register/DriverRegisterForm.tsx:24]: Move this list of static json values to "utils/FormValueData.json".

File: src/client/components/register/DriverRegisterForm.tsx [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+export default function DriverRegisterForm({
+  phoneNumber,
+  userRole,
+}: userDetailsFormProps) {
+  const { register, handleSubmit } = useForm<driverFormType>();
+  const navigate = useNavigate();
+  const { repo } = useRepository();
+  async function handleDriverDataSubmit(data: driverFormType) {
+    const formData = new FormData();
+    formData.append("firstName", data.firstName);
```
Comment (ravigrauniyar) [inline src/client/components/register/DriverRegisterForm.tsx:108]: Let's find a better way to append form data.

File: src/client/components/register/UserRegisterForm.tsx [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+import { useRepository } from "../../../hooks/CustomHook";
+
+type formValueData = {
+  labelText?: string;
+  id: "firstName" | "lastName" | "address" | "password" | "userImage";
+  value?: string | undefined;
+  type: string;
+  required: boolean;
+};
+const formValue: formValueData[] = [
```
Comment (ravigrauniyar) [inline src/client/components/register/UserRegisterForm.tsx:13]: Same issues as DriverRegisterForm should be fixed in this file also.

File: src/provider/RepositoryProvider.tsx [hunk 23]
```diff
@@ -1,32 +1,24 @@
...
-  const driverApi: IDriverApi = new DriverApi();
-  const driverRepo: IDriverRepo = new DriverRepo(driverApi);
-
-  const passengerApi: IPassengerApi = new PassengerApi();
-  const passengerRepo: IPassengerRepo = new PassengerRepo(passengerApi);
-
-  const userApi: IUserApi = new UserApi();
-  const userRepo: IUserRepo = new UserRepo(userApi);
-
+  const base_url: string = "http://localhost:8000/api/v1/";
```
Comment (ravigrauniyar) [inline src/provider/RepositoryProvider.tsx:16]: Import this url from .env file.

File: src/repositories/Repository.ts [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+      loginInfo,
+      undefined,
+      false
+    );
+    return this._getResponse(responseData);
+  }
+  async isPhoneExist(
+    phoneNumber: string
+  ): Promise<RegisterResponse | JsonAPIErrorResp | undefined> {
+    const responseData = await this._api.get<RegisterResponse>(
```
Comment (ravigrauniyar) [inline src/repositories/Repository.ts:53]: Each of the following distinct methods should have a distinct response type. For eg. Response type for `isPhoneExist` could be `boolean`.

**Comment by ravigrauniyar** [comment]
> _ No description provided. _

File: src/admin/components/dashboard/Dashboard_map.tsx (hunk 9 above)
Comment (Someone-rahul) [reply src/admin/components/dashboard/Dashboard_map.tsx:4]: i added alias, but its not working as expected

**Review by ravigrauniyar — APPROVED** [review]

**Review by ravigrauniyar — APPROVED** [review]

**Pull request merged**